### PR TITLE
Multinode support

### PR DESCRIPTION
**Pull request by samnordmann**
## What
Add support for multinode setting

The user must provide the environment variables
- `MASTER_ADDR` (optional on a single-node execution): name of the host participating that will play the role of server.
- `MASTER_PORT` (optional): port to be used. If not provided, the default value `c10d::TCPStoreOptions::kDefaultPort`(= 29500) will be used.

### Single node
```
mpirun -np 6 <PYTORCH_BUILD_DIR>/bin/nvfuser_tests --gtest_filter=NVFuserTest.FusionMultiGPU_CUDA
```

### Multi-node
```
mpirun -np 6 -H <node1>:3,<node2>:3 -x MASTER_ADDR=<node1> -x MASTER_PORT=12345 <PYTORCH_BUILD_DIR>/bin/nvfuser_tests --gtest_filter=NVFuserTest.FusionMultiGPU_CUDA
```

### CODE REVIEW

File: csrc/multidevice/communicator.cpp [hunk 1]
```diff
@@ -64,12 +66,35 @@ int parseEnv(
...
     if (!env) {
-      return 1;
+      return false;
     }
   }
   local_size = std::atoi(env);
 
-  return 0;
+  // retrieves master address
+  env = std::getenv("MASTER_ADDR");
```
Comment (zasdfgbnm) [inline csrc/multidevice/communicator.cpp:75]: Does MPI always provide this, even on a node not connected to any network?

File: csrc/multidevice/communicator.cpp [hunk 2]
```diff
@@ -102,16 +127,37 @@ c10::intrusive_ptr<c10d::Backend> createBackend(
...
+  }
 
   // creates the backend
   c10d::TCPStoreOptions store_opts;
-  store_opts.isServer = (rank_ == server_rank) ? true : false;
-  auto store = c10::make_intrusive<c10d::TCPStore>("localhost", store_opts);
+  {
+    char hostname[HOST_NAME_MAX]; // NOLINT (modernize-avoid-c-arrays)
+    gethostname(hostname, HOST_NAME_MAX);
+    struct hostent* master_host = gethostbyname(master_addr_.c_str());
```
Comment (zasdfgbnm) [inline csrc/multidevice/communicator.cpp:152]: Does this need any header?

File: csrc/multidevice/communicator.cpp [hunk 3]
```diff
@@ -102,16 +127,37 @@ c10::intrusive_ptr<c10d::Backend> createBackend(
...
+    // we define the server as the process at the master host with local rank 0
+    store_opts.isServer = !strcmp(master_host->h_name, hostname) &&
```
Comment (zasdfgbnm) [inline csrc/multidevice/communicator.cpp:154]: What happened if my `/etc/hostname` mismatch with DNS server?

File: csrc/multidevice/communicator.cpp (hunk 1 above)
Comment (samnordmann) [reply csrc/multidevice/communicator.cpp:75]: No, MPI never provides it, the user must provide it

File: csrc/multidevice/communicator.cpp (hunk 2 above)
Comment (samnordmann) [reply csrc/multidevice/communicator.cpp:152]: yes, `<netdb.h>`, which is included here through the inclusion chain, but let me include it explicitly

File: csrc/multidevice/communicator.cpp (hunk 3 above)
Comment (samnordmann) [reply csrc/multidevice/communicator.cpp]: Mhm, I am not certain, there could be an issue here indeed. I propose small revision where we pass `hostname` to `gethostbyname` also. This is the best possibility I can think about for now so let me know if you have a better suggestion.

File: csrc/multidevice/communicator.cpp (hunk 1 above)
Comment (samnordmann) [reply csrc/multidevice/communicator.cpp:76]: At some point, we would want to use python frontend anyway
https://github.com/NVIDIA/Fuser/issues/639

**Comment by samnordmann** [comment]
!build

File: csrc/multidevice/communicator.cpp [hunk 4]
```diff
@@ -64,12 +67,36 @@ int parseEnv(
...
+
+  // retrieves master port
+  env = std::getenv("MASTER_PORT");
+  if (env) {
+    master_port = std::atoi(env);
+  } else {
+    if (master_addr != "localhost") {
+      TORCH_WARN(
+          "the environment variable MASTER_PORT "
+          "has not been specified. Set to default");
```
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/multidevice/communicator.cpp:95]: Sorry, I realized it's in the PR description.

**Comment by samnordmann** [comment]
@zasdfgbnm with the last commit the master address defaults to "localhost" if there is only one node.
Lmk if the PR looks good now

**Comment by samnordmann** [comment]
!build

**Review by zasdfgbnm — APPROVED** [review]

File: csrc/multidevice/communicator.cpp (hunk 1 above)
Comment (IvanYashchuk) [reply csrc/multidevice/communicator.cpp:76]: MPI doesn't provide `MASTER_ADDR`, but you can query world size and world rank using `MPI_Comm_size` and `MPI_Comm_rank`. Why are these variables expected to be defined with environment variables?

Comment (samnordmann) [reply csrc/multidevice/communicator.cpp:76]: > MPI doesn't provide `MASTER_ADDR`

Correct. However a master address is needed to create a pytorch Store (necessary for Process Groups).
If you use torchrun though, `MASTER_ADDR` will be automatically set. 

> you can query world size and world rank using MPI_Comm_size and MPI_Comm_rank

We don't directly use OpenMPI library in nvFuser, we do everything through pytorch's process group. So the functions MPI_Comm_size or MPI_Comm_rank are not available.
The environment variables from which we retrieve the rank and size are always assumed to be set in MPI processes (also when using OpenMPI library), that is why we use it.

Comment (IvanYashchuk) [reply csrc/multidevice/communicator.cpp:76]: Thank you for the explanations!

**Pull request merged**